src/mimemail.erl

```diff
@@ -1143,10 +1150,13 @@ dkim_hash_data(CanonicHeaders, DkimHeader) ->
 	crypto:hash(sha256, <<JoinedHeaders/binary, DkimHeader/binary>>).
 
 dkim_sign(DataHash, {pem_plain, PrivBin}) ->
+	%% TODO: Conditionally set EcdhParams for public_key:sign
+	%% EcdhParams = crypto:ec_curve(secp521r1),
 	[PrivEntry] = public_key:pem_decode(PrivBin),
 	Key = public_key:pem_entry_decode(PrivEntry),
-	public_key:sign({digest, DataHash}, sha256, Key);
+	public_key:sign({digest, DataHash}, sha256, [Key, ed25519]);
```

Comment:
ok, it's actually more complex than I thought... Let me read-up for a bit...

Reply:
Looking at https://erlang.org/doc/man/public_key.html#sign-4 and https://erlang.org/doc/man/public_key.html#type-private_key, it seems just the fact that private key `Key` have been generated with `-algorithm ed25519` openssl switch, would make this function to sign with `ed25519` algorithm. May you try to just revert the change you made on this line?

Reply:
That change was tried as it didn't worked initially. So I don't really know the cause of the problem.

Reply:
Interesting... I see this is the commint where this algorithm have been introduced in OTP https://github.com/erlang/otp/commit/d743f51960384dd17ae3cb905ec57e8012c3abd7, but I didn't manage to make it work... I'll try to play with it later today...

```erlang
1> {ok, PrivBin} = file:read_file("private-key.pem").                             
{ok,<<"-----BEGIN PRIVATE KEY-----\nMC4CAQAwBQYDK2VwBCIEIBJoIIOo2SN5Wwtpf9seszQ4uCV+JF5MSW36oum6MlYH\n-----END PRIVAT"...>>}
2> [PrivEntry] = public_key:pem_decode(PrivBin).                                                   
[{'PrivateKeyInfo',<<48,46,2,1,0,48,5,6,3,43,101,112,4,34,                                       
                     4,32,18,104,32,131,168,217,35,121,91,                 
                     11,...>>,                   
                   not_encrypted}]        
3> Key = public_key:pem_entry_decode(PrivEntry).       
{'ECPrivateKey',1,
                <<4,32,18,104,32,131,168,217,35,121,91,11,105,127,219,30,
                  179,52,56,184,37,126,36,94,76,73,...>>,
                {namedCurve,{1,3,101,112}},
                asn1_NOVALUE,asn1_NOVALUE}
4> public_key:sign({digest, crypto:hash(sha256, <<"hello world">>)}, sha256, Key).           
** exception error: bad argument
     in function  crypto:pkey_sign_nif/5
        called as crypto:pkey_sign_nif(eddsa,sha256,
                                       {digest,<<185,77,39,185,147,77,62,8,165,46,82,215,218,
                                                 125,171,250,196,132,239,227,122,83,128,238,...>>},
                                       [<<4,32,18,104,32,131,168,217,35,121,91,11,105,127,219,30,
                                          179,52,56,184,37,126,36,94,...>>,
                                        ed25519],
                                       [])
     in call from crypto:sign/5 (crypto.erl, line 1366)
```

Reply:
Exactly the same problem I'm getting.

Reply:
Well, don't ask how, but I noticed that `public_key:sign/3` fails while calling this function deep in the crypto NIF code:
https://github.com/erlang/otp/blob/fc0e68ded3929c924f0e98315a48bc7e8763e3be/lib/crypto/c_src/eddsa.c#L52

and then I noticed that the `#'ECPrivateKey'.privateKey` is 34 bytes instead of 32 (as specified in the [RFC-8032](https://datatracker.ietf.org/doc/html/rfc8032#section-5.1.5)):
```erlang
> rr(public_key).
> {ok, PrivBin} = file:read_file("private-key.pem").
> [PrivEntry] = public_key:pem_decode(PrivBin).
> Key = public_key:pem_entry_decode(PrivEntry).
> size(Key#'ECPrivateKey'.privateKey).                                                      
3
```
then I tried to cut 1st 2 bytes of the key, and it somehow worked:
```erlang
> <<_:2/binary, PKey1/binary>> = Key#'ECPrivateKey'.privateKey.
> Key1 = Key#'ECPrivateKey'{privateKey = PKey1}.
> public_key:sign({digest, crypto:hash(sha256, <<"hello world">>)}, none, Key1).            
<<58,59,124,100,217,178,182,232,181,204,220,166,242,228,
  134,81,15,235,67,157,176,188,59,165,218,56,186,103,162,
  ...>>
```
I'm not yet sure it produces correct result though.

So far it looks like some bug in OTP private key decoder...

Reply:
Actually, if I try to parse "private-key.pub" using openssl, it produces the same 34 byte result as well!

```
$ openssl asn1parse -in private-key.pem
    0:d=0  hl=2 l=  46 cons: SEQUENCE          
    2:d=1  hl=2 l=   1 prim: INTEGER           :00
    5:d=1  hl=2 l=   5 cons: SEQUENCE          
    7:d=2  hl=2 l=   3 prim: OBJECT            :ED25519
   12:d=1  hl=2 l=  34 prim: OCTET STRING      [HEX DUMP]:04206D92AA87366D29DCFCA0D05930A22410C82059D7C21B3131B0A9E67272D735D0

1> byte_size(binary:decode_hex(<<"04206D92AA87366D29DCFCA0D05930A22410C82059D7C21B3131B0A9E67272D735D0">>)).
34
```

Reply:
Erlang's `crypto:generate_key(eddsa, ed25519).` produces 32-byte keys... And they use this `generate_key` in OTP tests. I think I'm going to report an issue in OTP and see what they say. I'm surprised opennssl's asn1parse produces 34 bytes as well...

Reply:
Ok, apparently someone already reported this: it's a known bug in OTP https://github.com/erlang/otp/issues/5156